### ISSUE TITLE
Matplotlib chooses the wrong font for unrecognized weights

### ISSUE THREAD

**Issue by Rufflewind** [matplotlib/matplotlib#8550]
### Bug report

**Bug summary**

With a lot of the modern fonts containing exotic font weights like “Thin” or “Ultra bold”, matplotlib will sometimes pick the wrong font weight because it assumes that any font weight that it doesn’t recognize is either normal (400) or bold (700), even though there is a “Regular” available.  Ultimately which one matplotlib picks *seems* to be actually kind of random, probably depending on the order in which the fonts appear in the font cache.

This is what the [logic in `font_manager.py`](https://github.com/matplotlib/matplotlib/blob/734c09f164819854a13d77b128da283952eb08d4/lib/matplotlib/font_manager.py#L438-L443) looks like:

```py
weight = next((w for w in weight_dict if sfnt4.find(w) >= 0), None)
if not weight:
    if font.style_flags & ft2font.BOLD:
        weight = 700
    else:
        weight = 400
```

`sfnt4` appears to be a string that contains the full name of the font, including the weight.  The problem with this approach is that if it doesn’t recognize the weight, it will simply assume the weight is either 400 or 700.  [A “Thin” font that was mistakenly labeled 400 might get picked over a “Regular” font.](http://stackoverflow.com/q/32091868)

Here are all the [weights currently recognized](https://github.com/matplotlib/matplotlib/blob/734c09f164819854a13d77b128da283952eb08d4/lib/matplotlib/font_manager.py#L95-L109):

```py
weight_dict = {
    'ultralight' : 100,
    'light'      : 200,
    'normal'     : 400,
    'regular'    : 400,
    'book'       : 400,
    'medium'     : 500,
    'roman'      : 500,
    'semibold'   : 600,
    'demibold'   : 600,
    'demi'       : 600,
    'bold'       : 700,
    'heavy'      : 800,
    'extra bold' : 800,
    'black'      : 900}
```

I imagine we could add “Thin” and some others into the list, but I don’t think this solves the fundamental problem that fonts with unrecognized weights are *presumed* “Regular”/“Bold” and do not have lower precedence than fonts that are *definitely* “Regular”/“Bold”.

TL;DR: It would be nice to have a more robust way to detect font weights and handle unknown font weights.

**Matplotlib version**

  * Operating System: Arch Linux
  * Matplotlib Version: HEAD
  * Python Version: 3.6.0
  * Jupyter Version (if applicable): n/a
  * Other Libraries: n/a

**Comment by Rufflewind** [comment]
This is a bit of "wish" but I think it would nice to have the option to specify the subfamily explicitly, in case the subfamily is really exotic.  So I should be able to pass e.g. `font="Fira Sans Four"` and get the `Four` subfamily.  (Either parse the subfamily from the user-provided string, or have a second attribute `font_subfamily`, which is arguably more principled.)

(@tacaswell)

**Comment by tacaswell** [comment]
Can you even do that with the font-config tools?  That seems like a difficult parsing problem as the space is used to both in the family name (which I think in 'Fire Sans') and to separate the sub-family.

There is a PR from @mdboom to replace our home-grown font selection with font-config around...

**Comment by Rufflewind** [comment]
Looks like font-config uses a colon:

```
$ fc-match 'Fira Sans'        
FiraSans-Regular.ttf: "Fira Sans" "Regular"
$ fc-match 'Fira Sans: Italic'
FiraSans-Italic.ttf: "Fira Sans" "Italic"
```

(I'm actually surprised that font-config even recognizes subfamily)

**Comment by anntzer** [comment]
#16203 fixes most of this, although I guess the "access by subfamily" request remains reasonable.
